### PR TITLE
rewrite gokeyless server certificate initialization

### PR DESCRIPTION
**Pull request by lziest**
- initializeServerCertAndKey focuses on getting cert and key issued,
  replacing initializeServer
- remove init-cert and init-key, we don't neet them
- don't write to token.json
- server restarted by SIGHUP will check server certificate and do auto renewal.

### CODE REVIEW

**Comment by lziest** [comment]
should address #121

File: cmd/gokeyless/gokeyless.go [hunk 1]
```diff
@@ -54,31 +55,32 @@ func init() {
...
 	flag.Parse()
 
-	s, err := server.NewServerFromFile(certFile, keyFile, caFile, net.JoinHostPort("", port), "")
-	if err != nil {
-		log.Warningf("Could not create server. Running initializeServer to get %s and %s", keyFile, certFile)
-
-		// If an existing CSR (and associated key) exists, don't
-		// proceed/overwrite the key.
-		if _, err := os.Stat(csrFile); !os.IsNotExist(err) {
-			log.Fatalf("an existing CSR was found at %q: remove once a signed certificate has been installed.", csrFile)
```
Comment (elithrar) [inline cmd/gokeyless/gokeyless.go]: I still want us to support the manual activation path as a backup/fallback if the automatic activation path fails. 

Having this guard prevents the end-user from running `$ gokeyless -manual-activation` again by accident, writing a new CSR to disk, and invalidating the one that might still be "in transit" (read: waiting to be signed + returned).

File: cmd/gokeyless/gokeyless.go [hunk 2]
```diff
@@ -54,31 +55,32 @@ func init() {
...
-		}
-
+	init := needNewCertAndKey()
+	if init {
+		log.Info("Now let's get server certificate and key")
 		// Allow manual activation (requires the CSR to be manually signed).
 		if manualMode {
+			log.Info("[Manual mode]")
```
Comment (elithrar) [inline cmd/gokeyless/gokeyless.go]: Likely redundant, if you passed `-manual-activation` to get here.

File: cmd/gokeyless/gokeyless.go [hunk 3]
```diff
@@ -54,31 +55,32 @@ func init() {
...
 			if err := manualActivation(); err != nil {
 				log.Fatal(err)
 			}
 
 			log.Infof("CSR generated (requires manual signing) and written to %q",
 				csrFile)
+			log.Info("We are done, please contact support for getting the CSR signed into a certificate")
```
Comment (elithrar) [inline cmd/gokeyless/gokeyless.go]: Redundant.

File: cmd/gokeyless/gokeyless.go [hunk 4]
```diff
@@ -54,31 +55,32 @@ func init() {
...
+	if init {
+		log.Info("Now let's get server certificate and key")
 		// Allow manual activation (requires the CSR to be manually signed).
 		if manualMode {
+			log.Info("[Manual mode]")
 			if err := manualActivation(); err != nil {
 				log.Fatal(err)
 			}
 
 			log.Infof("CSR generated (requires manual signing) and written to %q",
```
Comment (elithrar) [inline cmd/gokeyless/gokeyless.go]: Update this to encapsulate below: `"CSR generated (contact CloudFlare for manual signing) and written to %q`

File: cmd/gokeyless/gokeyless.go [hunk 5]
```diff
@@ -108,6 +110,23 @@ func main() {
 				log.Fatal(err)
 			}
 			s.Keys = keys
+
+			log.Info("Check server certificate...")
+			needRenewal := needNewCertAndKey()
```
Comment (elithrar) [inline cmd/gokeyless/gokeyless.go]:
```
if needNewCertAndKey() {
    ...
```

File: cmd/gokeyless/gokeyless.go [hunk 6]
```diff
@@ -122,6 +141,38 @@ func LoadKey(in []byte) (priv crypto.Signer, err error) {
 	return derhelpers.ParsePrivateKeyDER(in)
 }
 
+func validCert(cert *x509.Certificate) bool {
+	now := time.Now()
+	if now.Before(cert.NotBefore) {
```
Comment (elithrar) [inline cmd/gokeyless/gokeyless.go]:
```
if now.Before(cert.NotBefore) && now.After(cert.NotAfter) {
    return false
}
```

File: cmd/gokeyless/gokeyless.go [hunk 7]
```diff
@@ -122,6 +141,38 @@ func LoadKey(in []byte) (priv crypto.Signer, err error) {
...
+}
+func needNewCertAndKey() bool {
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		log.Errorf("cannot load server cert/key: %v", err)
+		return true
+	}
+
+	if !validCert(cert.Leaf) {
+		log.Errorf("certificate is either not yet valid, expired or expiring in a month")
```
Comment (elithrar) [inline cmd/gokeyless/gokeyless.go]: Can we be more precise and show the actual cause - not valid/expired is different from expiring in a month (and I'm not sure we should force failure if hasn't *actually* expired yet)

File: cmd/gokeyless/gokeyless.go (hunk 6 above)
Comment (lziest) [inline cmd/gokeyless/gokeyless.go]: This is incorrect, this should be `||`, I prefer not to changing it.

File: cmd/gokeyless/gokeyless.go (hunk 1 above)
Comment (lziest) [inline cmd/gokeyless/gokeyless.go]: thanks, I support this. Once a csr and key are generated, keep using it.

File: cmd/gokeyless/gokeyless.go (hunk 2 above)
Comment (lziest) [inline cmd/gokeyless/gokeyless.go]: removed.

File: cmd/gokeyless/gokeyless.go (hunk 4 above)
Comment (lziest) [inline cmd/gokeyless/gokeyless.go]: updated.

File: cmd/gokeyless/gokeyless.go (hunk 3 above)
Comment (lziest) [inline cmd/gokeyless/gokeyless.go]: removed.

File: cmd/gokeyless/gokeyless.go (hunk 5 above)
Comment (lziest) [inline cmd/gokeyless/gokeyless.go]: fixed.

File: cmd/gokeyless/gokeyless.go (hunk 7 above)
Comment (lziest) [inline cmd/gokeyless/gokeyless.go]: updated.

File: cmd/gokeyless/gokeyless.go (hunk 1 above)
Comment (lziest) [inline cmd/gokeyless/gokeyless.go]: The flag of `-manual-activation` now would check csr and key file, and matching them through public key. Also, we make it clear that `-manual-activation` won't start the server.

File: server/server.go [hunk 8]
```diff
@@ -85,7 +85,7 @@ func (keys *DefaultKeystore) LoadKeysFromDir(dir string, LoadKey func([]byte) (c
 		}
 
 		if !info.IsDir() && keyExt.MatchString(info.Name()) {
-			log.Debugf("Loading %s...\n", path)
+			log.Infof("Loading %s...\n", path)
```
Comment (kisom) [inline server/server.go]: I don't think the \n is needed here.

File: cmd/gokeyless/gokeyless.go [hunk 9]
```diff
@@ -35,16 +38,16 @@ var (
 	manualMode   bool
 )
 
+type CertError error
```
Comment (lziest) [inline cmd/gokeyless/gokeyless.go]: removed, it's leftover code.

File: server/server.go (hunk 8 above)
Comment (lziest) [inline server/server.go]: yeah, removed.

**Comment by elithrar** [comment]
Will give a final eyeball Friday morning PST.

**Comment by elithrar** [comment]
Approved.

**Pull request merged**